pkg/proxy/ipvs/netlink_linux.go

```diff
@@ -124,72 +126,41 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
-
-// If dev is specified, it's equivalent to exec:
-// $ ip route show table local type local proto kernel dev kube-ipvs0
-// 10.0.0.1  scope host  src 10.0.0.1
-// 10.0.0.10  scope host  src 10.0.0.10
-// Then cut the unique src IP fields,
-// --> result set: [10.0.0.1, 10.0.0.10]
-
-// If filterDev is specified, the result will discard route of specified device and cut src from other routes.
-func (h *netlinkHandle) GetLocalAddresses(dev, filterDev string) (sets.String, error) {
-	chosenLinkIndex, filterLinkIndex := -1, -1
-	if dev != "" {
-		link, err := h.LinkByName(dev)
-		if err != nil {
-			return nil, fmt.Errorf("error get device %s, err: %v", dev, err)
-		}
-		chosenLinkIndex = link.Attrs().Index
-	} else if filterDev != "" {
-		link, err := h.LinkByName(filterDev)
-		if err != nil {
-			return nil, fmt.Errorf("error get filter device %s, err: %v", filterDev, err)
-		}
-		filterLinkIndex = link.Attrs().Index
+// GetAllLocalAddresses return all local addresses on the node.
+// Only the addresses of the current family are returned.
+// IPv6 link-local addresses are excluded.
+func (h *netlinkHandle) GetAllLocalAddresses() (sets.String, error) {
+	addr, err := net.InterfaceAddrs()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get addresses: %v", err)
 	}
+	return utilproxy.AddressSet(h.isValidForSet, addr), nil
+}
 
-	routeFilter := &netlink.Route{
-		Table:    unix.RT_TABLE_LOCAL,
-		Type:     unix.RTN_LOCAL,
-		Protocol: unix.RTPROT_KERNEL,
+// GetLocalAddresses return all local addresses for an interface.
+// Only the addresses of the current family are returned.
+// IPv6 link-local addresses are excluded.
+func (h *netlinkHandle) GetLocalAddresses(dev string) (sets.String, error) {
+	ifi, err := net.InterfaceByName(dev)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get interface %s: %v", dev, err)
 	}
-	filterMask := netlink.RT_FILTER_TABLE | netlink.RT_FILTER_TYPE | netlink.RT_FILTER_PROTOCOL
+	addr, err := ifi.Addrs()
+	if err != nil {
+		return nil, fmt.Errorf("Can't get addresses from %s: %v", ifi.Name, err)
+	}
+	return utilproxy.AddressSet(h.isValidForSet, addr), nil
+}
 
-	// find chosen device
-	if chosenLinkIndex != -1 {
-		routeFilter.LinkIndex = chosenLinkIndex
-		filterMask |= netlink.RT_FILTER_OIF
+func (h *netlinkHandle) isValidForSet(ip net.IP) bool {
+	if h.isIPv6 != netutils.IsIPv6(ip) {
+		return false
 	}
-	routes, err := h.RouteListFiltered(netlink.FAMILY_ALL, routeFilter, filterMask)
-	if err != nil {
-		return nil, fmt.Errorf("error list route table, err: %v", err)
+	if h.isIPv6 && ip.IsLinkLocalUnicast() {
+		return false
 	}
-	res := sets.NewString()
-	for _, route := range routes {
-		if route.LinkIndex == filterLinkIndex {
-			continue
-		}
-		if h.isIPv6 {
-			if route.Dst.IP.To4() == nil && !route.Dst.IP.IsLinkLocalUnicast() {
-				res.Insert(route.Dst.IP.String())
-			}
-		} else if route.Src != nil {
-			res.Insert(route.Src.String())
-		}
+	if ip.IsLoopback() {
```

Comment:
should we mention in the comments that loopbacks are already excluded, it only mentions about ipv6 link-locals

Reply:
I mean in `pkg/proxy/ipvs/netlink.go`

Reply:
Updated